qmi/instruments/thorlabs/mpc320.py

```diff
@@ -0,0 +1,507 @@
+    MOT_SET_EEPROMPARAMS,
+    POL_GET_SET_PARAMS,
+)
+from qmi.instruments.thorlabs.apt_protocol import (
+    AptChannelJogDirection,
+    AptChannelState,
+    AptMessageId,
+    AptProtocol,
+)
+
+# Global variable holding the logger for this module.
+_logger = logging.getLogger(__name__)
+
+
+@dataclass
+class Thorlabs_MPC320_Status:
+    """
+    Data class for the status of the MPC320
+
+    Attributes:
+        channel:        Channel number.
+        position:       Absolute position of the channel in degrees.
+        velocity:       Velocity in controller units.
+        motor_current:  Current of motor in mA
+    """
+
+    channel: int
+    position: float
+    velocity: int
+    motor_current: int
+
+
+@dataclass
+class Thorlabs_MPC320_PolarisationParameters:
+    """
+    Data class for the polarisation parameters of the MPC320
+
+    Attributes:
+        velocity:       The velocity in percentage of the max velocity 400 degrees/s.
+        home_position:  The home position of all the paddles/channels in degrees.
+        jog_step1:      The position to move paddel/channel 1 by for a jog step in degrees.
+        jog_step2:      The position to move paddel/channel 2 by for a jog step in degrees.
+        jog_step3:      The position to move paddel/channel 3 by for a jog step in degrees.
+    """
+
+    velocity: float
+    home_position: float
+    jog_step1: float
+    jog_step2: float
+    jog_step3: float
+
+
+Thorlabs_MPC320_ChannelMap: Dict[int, int] = {1: 0x01, 2: 0x02, 3: 0x04}
+
+
+class Thorlabs_MPC320(QMI_Instrument):
+    """
+    Driver for a Thorlabs MPC320 motorised fibre polarisation controller.
+    """
+
+    DEFAULT_RESPONSE_TIMEOUT = 1.0
+
+    # the maximum range for a paddle is 170 degrees
+    # the value returned by the encoder is 1370 for 170 degrees
+    ENCODER_CONVERSION_UNIT = 170 / 1370
+
+    MIN_POSITION_DEGREES = 0
+    MAX_POSITION_DEGREES = 170
+
+    MIN_VELOCITY_PERC = 10
+    MAX_VELOCITY_PERC = 100
+
+    MIN_CHANNEL_NUMBER = 1
+    MAX_CHANNEL_NUMBER = 3
+
+    def __init__(self, context: QMI_Context, name: str, transport: str) -> None:
+        """Initialize the instrument driver.
+
+        Parameters:
+            name:       Name for this instrument instance.
+            transport:  QMI transport descriptor to connect to the instrument.
+        """
+        super().__init__(context, name)
+        self._transport = create_transport(transport, default_attributes={"baudrate": 115200, "rtscts": True})
+        self._apt_protocol = AptProtocol(self._transport, default_timeout=self.DEFAULT_RESPONSE_TIMEOUT)
+
+    def _validate_position(self, pos: float) -> None:
+        """
+        Validate the position. Any position for the MPC320 needs to be in the range 0 to 170 degrees,
+        or 0 to 1370 in encoder counts.
+
+        Parameters:
+            pos:    Position to validate in degrees.
+
+        Raises:
+            an instance of QMI_InstrumentException if the position is invalid.
```

Comment:
Minor: In the readthedocs this looks nicer if you do `QMI_InstrumentException: If the position is invalid`. The `:` makes sure the exception name is in bold and the description with regular lettering. Valid to all `Raises:` sections in these docstrings.

Reply:
nice! did it for all